### DISCUSSION TITLE
How to use this nlohmann library to create a dynamic JSON?

### DISCUSSION THREAD

**Discussion by jayjadhao** [nlohmann/json#2929] [Q&A] [answered]
I am using a macro NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE for creating a static JSON where its Object and Member value are fixed, eg: (Request JSON)
```json
{
        "allSampleParameters": 
        {
             "parameters": [
                        "92109",
                        "92110",
                        "92111",
                        ... ,
                        ... ,
                        ...
        ]
        }
}
```
As "parameters" is an array, this macro works well. As it can take many values in it. But when it comes to response to sending the key for each parameter (which is dynamic) eg: (Response JSON)
```json
{
"92109": "xyz",
"92110":"abc",
"92111":"efg",
... ,
... ,
...
}
```
How shall I use this macro with dynamic member value for defined objects? If it is not possible with macros, is their any other option?

**Comment by nlohmann** [comment, marked as answer]
Without knowing your type, i can't give you any hints.

**Reply by jayjadhao** [reply]
Sorry, I did not get you. Type means?

**Reply by nlohmann** [reply]
You neither shared the C++ code nor the error message you see. Which class are you using `NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE` with?

**Reply by jayjadhao** [reply]
Header file content (definition for both Request and Response)
Request.h :-
```cpp
                    #include <iostream>
                    #include <vector>
                    #include <string>
                    #include "../../nlohmann/json.hpp"
                    using namespace nlohmann;
                    using namespace std;
                     
                    namespace TestPayloadRequest {
                    
                        struct AllSampleParameters {
                            string allSampleParameters;
                            vector<string> parameters;
                        };
                        NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(AllSampleParameters, allSampleParameters, parameters)
                    }
```
Response.h:-
```cpp
                    #include <iostream>
                    #include <vector>
                    #include <string>
                    #include "../../nlohmann/json.hpp"
                    using namespace nlohmann;
                    using namespace std;
                     
                    namespace TestPayloadResponse {
                    
                        struct Parameters {
                            string 92109;
                            string 92110;
                            string 92111;
                        };
                    
                        struct ParameterKeyValue {
                            Parameters parameters;
                        };
                        NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(Parameters, 92109, 92110, 92111)
                        NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(ParameterKeyValue, parameters)
                    }
```

Testing the created data structure using the macros (library), test code:
```cpp
                    #include <iostream>
                    #include <string>
                    #include <gtest/gtest.h>
                    
                    #include "../../src/json-data-structures/include/Test_Request.h"
                    #include "../../src/json-data-structures/include/Test_Response.h"
                    #include "../../src/nlohmann/json.hpp"
                    
                    using namespace std;
                    using namespace nlohmann;
                    
                    TEST(TestPayloadRequest, CanConvertJsonTotestAllSampleParameters) 
                    {
	                    //Payload which is 
	                    auto jsonString = R"(
		                    {
			                    "allSampleParameters": "357186082267036",
			                    "parameters": 
			                    [
				                    "92109",
				                    "92110",
				                    "92111"
			                    ]
		                    }
	                    )"_json;
                    
                        TestPayloadRequest::AllSampleParameters allSampleParameters = jsonString;
                        assert(allSampleParameters == jsonString);
                    }
                    
                    TEST(TestPayloadResponse, CanConvertJsonTotestParameters) 
                    {
	                    //Payload depends on the request JSON string (the array content of "parameters" - which is dynamic)
	                    auto jsonString = R"(
		                    {
			                    "parameters": 
			                    {
				                    "92109": "102367",   // Key Value for each paramter request
				                    "92110": "198756",	 // Key Value for each paramter request
				                    "92111": "198567"	 // Key Value for each paramter request
			                    }
		                    }
	                    )"_json;
                    
                        TestPayloadResponse::ParameterKeyValue parameters = jsonString;
                        assert(parameters == jsonString);
                    }
```

I am looking for two issues:
1. If you can figure out, the "parameters" array can have many values in it and as per the response need to create with their respective key value. So the response is dynamic (member of the object).
2. Members of the object are numbers, and I am getting the below error:
         error: expected unqualified-id before numeric constant
         string 92109;

**Reply by nlohmann** [reply]
I do not understand the issue in 1.

(For 2.: a number is no valid variable name)

**Reply by jayjadhao** [reply]
For 1st:

The request payload and respective response can change as mentioned below:
Request:		
               {
			"allSampleParameters": "357186082267036",
			"parameters": 
			[
				"92109",
				"92110",
				"92111",
			]
		}
Respective Response:
               {
			"parameters": 
			{
				"92109": "102367",
				"92110": "198756",
				"92111": "198567"
                         }
              }
 
OR

Request:
		{
			"allSampleParameters": "357186082267036",
			"parameters": 
			[
				"92109",
				"92110",
				"92111",
                                "92112"
			]
		}
Respective Response:
               {
			"parameters": 
			{
				"92109": "102367",
				"92110": "198756",
				"92111": "198567"
                                "92112": "199039"
                         }
              }
OR

Request:
		{
			"allSampleParameters": "357186082267036",
			"parameters": 
			[
				"92109",
				"92110",
				"92111",
                                "92112",
                                "92113"
			]
		}
Respective Response:
               {
			"parameters": 
			{
				"92109": "102367",
				"92110": "198756",
				"92111": "198567"
                                "92112": "199039"
                                "92113": "193458"
                         }
              }
and so on.. (100 different combinations)
For each request, the Response Payload will change, as it needs to send a key for each parameters. Which is dynamic in nature, changes as per the received array value in the Request payload.

For 2nd, in this case, I have a "number" variable name.. what I can do?

**Reply by nlohmann** [reply]
1. So you have different JSON types, but want to create the same C++ type? Then you need to define a user-defined `from_json` function.
2. `string 92109;` is invalid C++. You need to add some prefix like `string v92109;`

**Reply by gregmarr** [reply]
You should be able to have `parameters` in the request be a `std::vector<std::string>` and `parameters` in the response be a `std::map<std::string, std::string>` and it should serialize as you provided.

**Reply by jayjadhao** [reply]
@gregmarr Okay. But what shall I add in the macro (which expect member value), NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(Parameters, 92109, 92110, 92111)

**Reply by jayjadhao** [reply]
@nlohmann Yes, C++ type. So do you mean, I can't use macro "NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE" ?

**Reply by gregmarr** [reply]
Just use this:

                        struct ParameterKeyValue {
                            std::map<std::string, std::string> parameters;
                        };
                        NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(ParameterKeyValue, parameters)

**Reply by jayjadhao** [reply]
Awesome. Really appriciate! It is working. Thank you so much @gregmarr and @nlohmann for the help.

**Reply by gregmarr** [reply]
You're welcome.
